space-before-function-parentheses errors on generator function expressions

```js
// test.js
function* foo () {}
var bar = function* () {};
```

```json
// config.json
{
	"ecmaFeatures": {
		"generators": true
	},
	"rules": {
		"generator-star": [2, "start"],
		"space-before-function-parentheses": 2
	}
}
```

```sh
$ eslint --reset --no-eslintrc --config config.json test.js

test.js
  3:18  error  Missing space before function parentheses  space-before-function-parentheses

✖ 1 problem (1 error, 0 warnings)
```

Moving this from #1920, as it's a separate issue. I propose that `space-before-function-parentheses` should not warn on generator function expressions at all, instead deferring to `generator-star`.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/9171102-space-before-function-parentheses-errors-on-generator-function-expressions?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/lib/rules/space-before-function-parentheses.js
+++ b/lib/rules/space-before-function-parentheses.js
@@ -59,13 +59,22 @@ module.exports = function(context) {
      * @returns {void}
      */
     function validateSpacingBeforeParentheses(node) {
-        var tokens = context.getTokens(node),
-            isNamed = isNamedFunction(node),
+        var isNamed = isNamedFunction(node),
+            tokens,
             leftToken,
             rightToken,
             location;
 
-        if (isNamed) {
+        if (node.generator && !isNamed) {
+            return;
+        }
+
+        tokens = context.getTokens(node);
+
+        if (node.generator) {
+            leftToken = tokens[2];
+            rightToken = tokens[3];
+        } else if (isNamed) {
             if (node.id) {
                 leftToken = tokens[1];
                 rightToken = tokens[2];

--- a/tests/lib/rules/space-before-function-parentheses.js
+++ b/tests/lib/rules/space-before-function-parentheses.js
@@ -28,6 +28,8 @@ eslintTester.addRuleTest("lib/rules/space-before-function-parentheses", {
             code: "var obj = { foo () {} };",
             ecmaFeatures: { objectLiteralShorthandMethods: true }
         },
+        { code: "function* foo () {}", ecmaFeatures: { generators: true } },
+        { code: "var foo = function *() {};", ecmaFeatures: { generators: true } },
 
         { code: "function foo() {}", options: ["never"] },
         { code: "var foo = function() {}", options: ["never"] },
@@ -38,24 +40,44 @@ eslintTester.addRuleTest("lib/rules/space-before-function-parentheses", {
             options: ["never"],
             ecmaFeatures: { objectLiteralShorthandMethods: true }
         },
+        {
+            code: "function* foo() {}",
+            options: ["never"],
+            ecmaFeatures: { generators: true }
+        },
+        {
+            code: "var foo = function*() {};",
+            options: ["never"],
+            ecmaFeatures: { generators: true }
+        },
 
         {
             code: [
                 "function foo() {}",
                 "var bar = function () {}",
+                "function* baz() {}",
+                "var bat = function*() {};",
                 "var obj = { get foo() {}, set foo(val) {}, bar() {} };"
             ].join("\n"),
             options: [ { named: "never", anonymous: "always" } ],
-            ecmaFeatures: { objectLiteralShorthandMethods: true }
+            ecmaFeatures: {
+                generators: true,
+                objectLiteralShorthandMethods: true
+            }
         },
         {
             code: [
                 "function foo () {}",
                 "var bar = function() {}",
+                "function* baz () {}",
+                "var bat = function* () {};",
                 "var obj = { get foo () {}, set foo (val) {}, bar () {} };"
             ].join("\n"),
             options: [ { named: "always", anonymous: "never" } ],
-            ecmaFeatures: { objectLiteralShorthandMethods: true }
+            ecmaFeatures: {
+                generators: true,
+                objectLiteralShorthandMethods: true
+            }
         }
     ],
 
@@ -119,6 +141,18 @@ eslintTester.addRuleTest("lib/rules/space-before-function-parentheses", {
                     message: "Missing space before function parentheses.",
                     line: 1,
                     column: 15
+                }
+            ]
+        },
+        {
+            code: "function* foo() {}",
+            ecmaFeatures: { generators: true },
+            errors: [
+                {
+                    type: "FunctionDeclaration",
+                    message: "Missing space before function parentheses.",
+                    line: 1,
+                    column: 13
                 }
             ]
         },
@@ -187,6 +221,19 @@ eslintTester.addRuleTest("lib/rules/space-before-function-parentheses", {
                     message: "Unexpected space before function parentheses.",
                     line: 1,
                     column: 15
+                }
+            ]
+        },
+        {
+            code: "function* foo () {}",
+            options: ["never"],
+            ecmaFeatures: { generators: true },
+            errors: [
+                {
+                    type: "FunctionDeclaration",
+                    message: "Unexpected space before function parentheses.",
+                    line: 1,
+                    column: 13
                 }
             ]
         },

